crates/transpiler/src/passes/remove_diagonal_gates_before_measure.rs

```diff
@@ -46,44 +51,61 @@ pub fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()>
     ];
     static DIAGONAL_3Q_GATES: [StandardGate; 1] = [StandardGate::CCZ];
 
-    let mut nodes_to_remove = Vec::new();
-    for (index, inst) in dag.op_nodes(true) {
-        if inst.op.name() == "measure" {
+    let run_in_parallel = getenv_use_multiple_threads();
+
+    let process_node = |index: NodeIndex, inst: &PackedInstruction| {
+        if matches!(
+            inst.op.view(),
+            OperationRef::StandardInstruction(StandardInstruction::Measure)
+        ) {
             let predecessor = (dag.quantum_predecessors(index))
                 .next()
                 .expect("index is an operation node, so it must have a predecessor.");
 
-            match &dag[predecessor] {
-                NodeType::Operation(pred_inst) => match pred_inst.standard_gate() {
-                    Some(gate) => {
-                        if DIAGONAL_1Q_GATES.contains(&gate) {
-                            nodes_to_remove.push(predecessor);
-                        } else if DIAGONAL_2Q_GATES.contains(&gate)
-                            || DIAGONAL_3Q_GATES.contains(&gate)
-                        {
-                            let mut successors = dag.quantum_successors(predecessor);
-                            if successors.all(|s| {
-                                let node_s = &dag.dag()[s];
-                                if let NodeType::Operation(inst_s) = node_s {
-                                    inst_s.op.name() == "measure"
-                                } else {
-                                    false
-                                }
-                            }) {
-                                nodes_to_remove.push(predecessor);
-                            }
+            let NodeType::Operation(ref pred_inst) = dag[predecessor] else {
+                return None;
+            };
+            if let Some(gate) = pred_inst.standard_gate() {
+                if DIAGONAL_1Q_GATES.contains(&gate) {
```

Comment:
I'm thinking using `matches!(gate,  ...)` here instead of `contains()` over a static array will end up with a more efficient check at the cost of more verbose code. I'll benchmark pivot the logic here to use that instead if it's measurable.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/14723/commits/6c9d71c7326087a2a6c94a4a61164328fb21e060